refactor: ♻️  user `Properties` classes in user facing functions

signekb opened the pull request:
## Description

This PR changes the input type (and output type, when it's relevant) to a `Properties` class in the user facing functions.

This means that the functions take a `Properties` class as input, but transforms it to a `dict` before handling it to the `checks`. Before returning, the `properties` object is transformed back to a `Properties` class using a new method `from_dict` based on the work by @martonvago. For this method, we needed a new dependency (`dacite`).

Closes #970 

<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [X] Added or updated tests
- [X] Updated documentation
- [X] Ran `just run-all`

signekb committed:
feat: :sparkles: add `from_dict` method to `Properties`

Based on the work by @martonvago in https://github.com/seedcase-project/seedcase-sprout/commit/498e1b977e9e2ba548aceb6fdaf46cdf632ccd76.

signekb committed:
Merge branch 'main' into refactor/use-properties-classes-in-userfacing-functions

signekb committed:
refactor: :recycle: change input type to be `Properties` class and transform to `dict` internally

signekb committed:
test: :white_check_mark: update tests with new input and output types

signekb committed:
docs: :memo: remove `compact_dict` from package guide

signekb commented on tests/core/test_create_resource_properties.py:
```diff
@@ -1,40 +1,44 @@
     """Given valid inputs, should create properties object correctly."""
     resource_path = tmp_path / "resources" / "1"
     resource_path.mkdir(parents=True)
-    properties = {
-        "name": "test",
-        "path": "",
-    }
```
Added a fixture, `resource_properties()`, instead of defining identical `properties` here and in L28-31.

signekb commented on tests/core/test_edit_package_properties.py:
```diff
@@ -48,20 +48,22 @@ def test_edits_only_changed_package_properties(properties_path, properties):
     # When, Then
-    expected_properties = current_properties | properties
-    assert edit_package_properties(properties_path, properties) == expected_properties
+    expected_properties = current_properties | properties.compact_dict
```
The easiest way to "merge" the properties are still by using the `dict` representation. So, that's why I compare `dicts` here.

signekb commented on tests/core/test_edit_package_properties.py:
```diff
@@ -83,22 +85,19 @@ def test_throws_error_if_current_package_properties_are_malformed(tmp_path, prop
+    """Should not accept custom fields to properties."""
     # Given
-    current_properties = read_json(properties_path)
-    new_properties = {"custom-field": "custom-value"}
+    new_properties = PackageProperties.from_dict({"custom-field": "custom-value"})
 
     # When, Then
-    assert (
-        edit_package_properties(properties_path, new_properties)
-        == current_properties | new_properties
-    )
+    with raises(NotPropertiesError):
+        assert edit_package_properties(properties_path, new_properties)
```
Note : This is a change of behaviour. 

When we used a dict, it was easy to add a custom field (as allowed by the Data Package standard). This, however, becomes tricker when we use the `Properties` classes. 

Are we fine with not supporting custom fields for now and then add it later, if we ever need/want to support them?

lwjohnst86 replied:
I'm fine with that. If we need to add custom fields, I think there's a section in the data package spec that talks about adding extensions, which we could do.

signekb commented on tests/core/test_properties.py:
```diff
@@ -99,3 +99,41 @@ def test_creates_package_properties_with_correct_defaults(mock_uuid):
+
+
+@mark.parametrize(
+    "dict, expected_properties",
```
Is there a more elegant way of testing this method for all `Properties` classes or is this fine?

lwjohnst86 replied:
Not sure, but my feeling is no :laughing: This is fine though, it is just test code.

martonvago replied:
I also think it's fine. I think the nested tests are very useful, the others are basically testing that all subclasses inherit the superclass method.

lwjohnst86 reviewed (dismissed):
Nice! Looks good :relaxed:

martonvago commented on tests/core/test_edit_package_properties.py:
```diff
@@ -83,22 +85,19 @@ def test_throws_error_if_current_package_properties_are_malformed(tmp_path, prop
 
 
-def test_adds_custom_fields(
+def test_throws_error_when_adding_custom_fields(
```
My only problem with this test is that what this is actually testing is that an empty properties object will lead to an error (so, the next test). This is because `from_dict` ignores unknown key-value pairs, so here it ignores all key-value pairs. But if we added a known key-value pair, `from_dict` would create an acceptable `PackageProperties` and no error would be raised.

Now that the input to the function is a `PackageProperties` (which cannot contain custom fields), I think we can just remove this test.

signekb replied:
You're absolutely right! :star2: I'll remove it.

lwjohnst86 replied:
@signekb when you do this, I can approve and merge :relaxed:

signekb replied:
Done 🔥

martonvago reviewed (changes requested):
Yay! :rocket: 

I will use this as a base for https://github.com/seedcase-project/seedcase-sprout/pull/974

signekb committed:
test: :fire: remove custom field test

Actually testing the same as `test_throws_error_if_new_properties_are_empty()`, since a custom field is removed anyway.

lwjohnst86 committed:
Merge branch 'main' into refactor/use-properties-classes-in-userfacing-functions

lwjohnst86 committed:
build: :arrow_up: regenerate lockfile
